hw/chisel/src/main/scala/snax/xdma/xdmaStreamer/AddressGenUnit.scala

```diff
@@ -38,6 +46,49 @@ class BasicCounter(width: Int, hasCeil: Boolean = true)
   }
 }
 
+class ProgrammableCounter(width: Int, hasCeil: Boolean = true)
+    extends Module
+    with RequireAsyncReset {
+  val io = IO(new Bundle {
+    val tick = Input(Bool())
+    val reset = Input(Bool())
+    val ceil = Input(UInt(width.W))
+    val step = Input(UInt((width - 1).W))
+
+    val value = Output(UInt(width.W))
+    val lastVal = Output(Bool())
+  })
+  val nextValue = Wire(UInt(width.W))
+  val value = RegNext(nextValue, 0.U)
+
+  // If has ceil, a small counter is used to count the step
+  // The small counter's function is to determine whether the ceil is reached, and a reset is needed.
+  if (hasCeil) {
+    val smallCounter = Module(new BasicCounter(width, hasCeil) {
+      override val desiredName = "ProgrammableCounter_SmallCounter"
+    })
+
+    smallCounter.io.tick := io.tick
+    smallCounter.io.reset := io.reset
+    smallCounter.io.ceil := io.ceil
+    io.lastVal := smallCounter.io.lastVal
+  }
+
+  nextValue := {
+    if (hasCeil) {
+      Mux(
+        io.reset || (io.lastVal && io.tick),
+        0.U,
+        Mux(io.tick, value + io.step, value)
+      )
+    } else {
+      Mux(io.reset, 0.U, Mux(io.tick, value + io.step, value))
+    }
+  }
+
+  io.value := value
+}
```

Comment:
Chisel is based on Scala. So, when writing Chisel, you are actually writing two languages simultaneously! I think this view is very important for people writing Chisel. 

For your question, **if / else** is *scala* condition, they are evaluated during the generation phase. **when / otherwise** is *chisel* condition, and they become the actual hardware implementation later. 

Other examples are **switch...case...** in scala and **case...is...** in chisel, **!=** in scala and **=/=** in chisel etc.

Reply:
Aha! Cool okay, I understand! Thanks 😄